pkg/controller/direct/alloydb/cluster_controller.go

```diff
@@ -0,0 +1,301 @@
+		return false, fmt.Errorf("cluster deletion failed: %w", err)
+	}
+	return true, nil
+}
+
+func (a *clusterAdapter) waitForOp(ctx context.Context, op *api.Operation) error {
+	for {
+		current, err := a.client.Projects.Locations.Operations.Get(op.Name).Context(ctx).Do()
+		if err != nil {
+			return fmt.Errorf("getting operation status of %q: %w", op.Name, err)
+		}
+		if current.Done {
+			if current.Error != nil {
+				return fmt.Errorf("operation %q completed with error: %v", op.Name, current.Error)
+			} else {
+				return nil
+			}
+		}
+		time.Sleep(2 * time.Second)
+	}
+}
+
+// Create implements the Adapter interface.
+func (a *clusterAdapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx)
+	log.V(0).Info("creating object", "u", u)
+
+	parent := "projects/" + a.projectID + "/locations/" + a.location
+	cluster := a.desired
+
+	log.V(0).Info("creating cluster", "cluster", cluster)
+	var op *api.Operation
+	var err error
+	// Default ClusterType to be PRIMARY
+	if cluster.ClusterType == "SECONDARY" {
+		op, err = a.client.Projects.Locations.Clusters.Createsecondary(parent, cluster).ClusterId(a.resourceID).Context(ctx).Do()
+	} else {
+		op, err = a.client.Projects.Locations.Clusters.Create(parent, cluster).ClusterId(a.resourceID).Context(ctx).Do()
+	}
+	if err != nil {
+		return fmt.Errorf("creating cluster: %w", err)
+	}
+
+	if err := a.waitForOp(ctx, op); err != nil {
+		return fmt.Errorf("waiting for cluster create %s: %w", a.fullyQualifiedName(), err)
+	}
+
+	created, err := a.client.Projects.Locations.Clusters.Get(a.fullyQualifiedName()).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("getting created cluster: %w", err)
+	}
+
+	log.V(0).Info("created cluster", "cluster", created)
+
+	resourceID := created.Name
+	if err := unstructured.SetNestedField(u.Object, resourceID, "spec", "resourceID"); err != nil {
+		return fmt.Errorf("setting spec.resourceID: %w", err)
+	}
+
+	mapCtx := &MapContext{
+		// kube: kube,
+	}
+	observedState := ClusterStatusFromApi(mapCtx, created)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return setStatus(u, observedState)
+}
+
+// Update implements the Adapter interface.
+func (a *clusterAdapter) Update(ctx context.Context, u *unstructured.Unstructured) error {
+	log := klog.FromContext(ctx)
+	log.V(0).Info("updating object", "u", u)
+	var latest *api.Cluster
+
+	updateMask := &fieldmaskpb.FieldMask{}
+	if a.desired.DisplayName != a.actual.DisplayName {
+		log.V(0).Info("change detected: displayName")
+		updateMask.Paths = append(updateMask.Paths, "displayName")
+	}
+	if a.desired.ClusterType != a.actual.ClusterType {
+		log.V(0).Info("change detected: clusterType")
+		updateMask.Paths = append(updateMask.Paths, "clusterType")
+	}
+	if !reflect.DeepEqual(a.desired.EncryptionConfig, a.actual.EncryptionConfig) {
+		log.V(0).Info("change detected: encryption_config")
+		updateMask.Paths = append(updateMask.Paths, "encryption_config")
```

Comment:
BTW, I'm trying to figure out if we've documented anywhere what our behaviour should be here.  mockgcp uses grpc-gateway, which is very strict about the names, but I think real-gcp allows either proto or json format.  If you find any docs, please LMK!